spatial-tracking-explainer.md

```diff
@@ -0,0 +1,382 @@
+
+```js
+let xrSession = null;
+let xrFrameOfReference = null;
+
+function onSessionStarted(session) {
+  xrSession = session;
+  xrSession.requestFrameOfReference({ type:'bounded' })
+  .then((frameOfReference) => {
+    xrFrameOfReference = frameOfReference;
+  })
+  .then(setupWebGLLayer)
+  .then(() => {
+    xrSession.requestAnimationFrame(onDrawFrame);
+  });
+}
+```
+
+The `XRBoundedFrameOfReference` also reports geometry within which the application should try to ensure that all content the user needs to interact with can be reached. This polygonal boundary represents a loop of points at the edges of the safe space. The points are given in a clockwise order as viewed from above, looking towards the negative end of the Y axis. The shape it describes is not guaranteed to be convex. The values reported are relative to the frame of reference origin, and must have a `y` value of `0` and a `w` value of `1`.
+
+```js
+// Demonstrated here using a fictional 3D library to simplify the example code.
+function createBoundsMesh() {
+  boundsMesh.clear();
+  
+  // Visualize the bounds geometry as 2 meter high quads
+  let pointCount = xrFrameOfReference.boundsGeometry.length;
+  for (let i = 0; i < pointCount - 1; ++i) {
+    let pointA = xrFrameOfReference.boundsGeometry[i];
+    let pointB = xrFrameOfReference.boundsGeometry[i+1];
+    boundsMesh.addQuad(
+        pointA.x, 0, pointA.z, // Quad Corner 1
+        pointB.x, 2.0, pointB.z) // Quad Corner 2
+  }
+  // Close the loop
+  let pointA = xrFrameOfReference.boundsGeometry[pointCount-1];
+  let pointB = xrFrameOfReference.boundsGeometry[0];
+  boundsMesh.addQuad(
+        pointA.x, 0, pointA.z, // Quad Corner 1
+        pointB.x, 2.0, pointB.z) // Quad Corner 2
+  }
+}
+```
+
+### Unbounded Frame of Reference
+A _unbounded_ experience is one in which the user is able to freely move around their physical environment. An example of this sort of experience would be a campus tour or a whole-house renovation preview. These experiences explicitly require that the user be unbounded in their ability to walk around, and the unbounded frame of reference will adjust its origin as needed to maintain optimal stability for the user, even if the user walks many meters from the origin. In doing so, the origin may drift from its original physical location.
```

Comment:
I would expect that large-scale, untethered VR experiences (like those that [The VOID](https://www.thevoid.com/) does) could be reasonably handled with an `unbounded` FoR. It's definitely easier to think about in terms of AR, though.

Reply:
See the discussion in https://github.com/immersive-web/anchors/issues/8#issuecomment-427473938.  As shipping today, developers doing desktop VR with Windows MR headsets with inside-out tracking (same as HoloLens) can benefit from independently-adjusting free-space anchors when their users are using backpack PCs to do `"unbounded"` experiences.

Reply:
That said, that's more an argument not to bake "AR" as a term into the name of the frame, moreso than a reason to avoid mentioning AR examples in prose for clarity.  It seems perfectly fine to mention ARCore phones and HoloLens as evocative examples of devices/platforms where users can wander for many 10s/100s of meters.

Reply:
I don't have a strong preference here but my inclination is that more use-cases and examples help clarify these sorts of API decisions - I know they help me a lot!

Reply:
When designing APIs, we are all often compelled to be more forward-looking than appears required by current hardware, which can sometimes lead to confusion for today's developers targeting today's hardware.

I've generally found documentation to be a great release valve for that tension.  We can be very specific to today's hardware in how we describe things in docs, since we can trivially update that doc prose later as the hardware landscape changes - we can't make that kind of freeform change to the API shape itself.

Reply:
I tried to strike a balance and have added a few examples to each section.  I'm wondering if it would also be helpful to make a flow chart that developers can use to determine which Frame Of Reference to use?  Maybe someone with better art skills than me wants to do that?

Reply:
Yeah, you mentioned the flow-chart idea when we discussed this and it helped me understand what you were going for in terms of the decision-making required by the app developer. +1 to creating the actual flowchart as an image - my art-fu is weak otherwise I would volunteer.